core/src/tests/dummy_transport.rs

```diff
@@ -85,14 +91,31 @@ impl Transport for DummyTransport {
         }
     }
 
-    fn dial(self, _addr: Multiaddr) -> Result<Self::Dial, (Self, Multiaddr)>
+    fn dial(self, addr: Multiaddr) -> Result<Self::Dial, (Self, Multiaddr)>
     where
         Self: Sized,
     {
-        unimplemented!();
+        let peer_id = if let Some(peer_id) = self.next_peer_id {
+            peer_id
+        } else {
+            PublicKey::Rsa((0 .. 128).map(|_| -> u8 { 1 }).collect()).into_peer_id()
+        };
+
+        let fut =
+            if addr.to_string() == "/unix/unreachable" {
+                let err_string = format!("unreachable host error, peer={:?}", peer_id);
+                future::err(io::Error::new(io::ErrorKind::Other, err_string))
+            } else {
+                future::ok((peer_id, DummyMuxer::new()))
+            };
+        Ok(Box::new(fut))
     }
 
-    fn nat_traversal(&self, _server: &Multiaddr, _observed: &Multiaddr) -> Option<Multiaddr> {
-        unimplemented!();
+    fn nat_traversal(&self, server: &Multiaddr, observed: &Multiaddr) -> Option<Multiaddr> {
+        if server == observed {
+            Some(observed.encapsulate("/udt").unwrap())
```

Comment:
I know this is a dummy, but this change doesn't really make sense.

Reply:
Yeah, it smells for sure. So the context is [this test](https://github.com/libp2p/rust-libp2p/pull/602/files#diff-0c5110fdcca8841806bc220275bed002R1268) where I test that "nat_traversal_transforms_the_observed_address_according_to_the_transport_used", so I need it to perform some kind of transform to assert on. While I agree that tossing on `/udt` is silly, I'm not sure what kind of transform a `DummyTransport` should be performing. I could just toss back the `observed` but that's also a little unsatisfying.
Do you have an idea for what is better here? pop off the port and, dunno, increment it by one?

Reply:
I changed this to act more like other `nat_traversal` impls and made it increment the port by one. Maybe slightly less nonsensical. :/